app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,598 @@
+  private ArrayList<FileItem> filesToSend = new ArrayList<>();
+  private FileListAdapter fileListAdapter;
+
+  private List<WifiP2pDevice> peerDevices = new ArrayList<WifiP2pDevice>();
+  private InetAddress selectedPeerDeviceInetAddress;
+  private InetAddress fileReceiverDeviceAddress;  // IP address of the file receiving device
+  private boolean fileTransferStarted = false;
+
+  private PeerGroupHandshakeAsyncTask peerGroupHandshakeAsyncTask;
+  private SenderDeviceAsyncTask[] senderDeviceAsyncTaskArray;
+  private ReceiverDeviceAsyncTask receiverDeviceAsyncTask;
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    KiwixApplication.getApplicationComponent().activityComponent()
+        .activity(this)
+        .build()
+        .inject(this);
+    setTheme(sharedPreferenceUtil.nightMode() ? R.style.Theme_AppCompat_DayNight_NoActionBar : R.style.AppTheme);
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_local_file_transfer);
+    ButterKnife.bind(this);
+
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    Intent filesIntent = getIntent();
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
+    if (fileUriArrayList != null && fileUriArrayList.size() > 0) {
+      setDeviceAsFileSender();
+    }
+
+    setSupportActionBar(actionBar);
+    actionBar.setNavigationIcon(R.drawable.ic_close_white_24dp);
+    actionBar.setNavigationOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        wifiDirectManager.closeLocalFileTransferActivity();
+      }
+    });
+
+    wifiDirectManager.initialiseWifiDirectManager();
+
+    listViewPeerDevices.setAdapter(
+        new WifiPeerListAdapter(this, R.layout.row_peer_device, peerDevices));
+
+    if (fileSendingDevice) {
+      totalFilesForTransfer = fileUriArrayList.size();
+
+      for (int i = 0; i < fileUriArrayList.size(); i++) {
+        filesToSend.add(new FileItem(getFileName(fileUriArrayList.get(i)), TO_BE_SENT));
+      }
+
+      displayFileTransferProgress();
+    }
+  }
+
+  @OnItemClick(R.id.list_peer_devices)
+  void onItemClick(int position) {
+    /* Connection can only be initiated by user of the sender device, & only when transfer has not been started */
+    if (!fileSendingDevice || fileTransferStarted) {
+      return;
+    }
+
+    WifiP2pDevice senderSelectedPeerDevice =
+        (WifiP2pDevice) listViewPeerDevices.getAdapter().getItem(position);
+    wifiDirectManager.setSenderSelectedPeerDevice(senderSelectedPeerDevice);
+    alertDialogShower.show(
```

Comment:
alertDialogShower can be injected in WifiDirectManager and you can create a callback `onHandshakeStarted` so this method can be reduced to `wifiDirectManager.sendToDevice(selectedDevice)`

Reply:
- I'm not clear on why you have suggested to create a `onHandshakeStarted` callback

- And I also have a query which is bothering me: Why do we use callbacks (i.e. an interface) instead of public methods?
Is it to enforce that a certain method/s (those specified in the callback) always exist when using a certain helper class (for greater modularity)?

Reply:
Short Answer: It is the [I](https://en.wikipedia.org/wiki/Interface_segregation_principle) in [SOLID](https://en.wikipedia.org/wiki/SOLID)

Long Answer: The less public methods the better, the more you expose the functionality of your class the less contained it's own functionality is, the more confusing it is to use and harder to change the functionality of.
If wifiDirectManager has 3 public methods `start/stop/send` then it is very easy for somebody else to come along and use it well (even better if these are defined in an interface) but if it has 15 then I _need_ to know how your class functions internally. You are writing code for another human to understand the first time they read it, not for the compiler

Reply:
`onHandshakeStarted` was just a suggestion as it is a point where you may want to perform a UI action, here it is only showing a toast but it is conceivable ui could be updated, toasts are weird because of their transient UI nature and can easily be shown in any piece of code by retrieving the activity instance, if you ever want to convert it to the cooler SnackBar you will neeed a view reference to show it and that would necessitate being in the activity

Reply:
Thanks for SOLID and the explanation. Previously my view of readability was restricted to self-explanatory names and comments. I now realise that I also need to look at my helper classes as some type of API - only methods relevant to the client should be made available to them.

I've [shifted to using AlertDialogShower](https://github.com/kiwix/kiwix-android/commit/083d7768b8fdf58803fc138bf0ab6c35be32f432#diff-7e01f6f425daaca67175b74c839c6af2R60), but I suspect injecting WifiDirectManager will lead to changes.

Will get to converting the methods to callbacks

Reply:
Injecting WifiDirectManager should be easy as I already provided the injection for DialogShower (note: you inject the interface and dagger will provide the impl, which is sweet).
By just annotating a primary public constructor with `@Inject` if dagger can reason about all the parameters already it will provide it without you creating a specific binding

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/72ccd4f7e3f5f31586df40211dc2d7507e827c2d#diff-7e01f6f425daaca67175b74c839c6af2R83)